src/main/java/com/example/petbuddybackend/service/chat/session/ChatSessionService.java

```diff
@@ -2,101 +2,190 @@
-        ChatUserMetadata metadata = new ChatUserMetadata(username, sessionId, timeZone);
-        chatSessionManager.put(chatId, metadata);
+    /**
+     * @return Map with chatId as key and number of sessions as value
+     * */
+    private Map<Long, Integer> countChatRoomSessions(String username) {
+        Set<SimpSession> userSessions = webSocketSessionService.getUserSessions(username);
+        Map<Long, Integer> chatRoomSessions = new HashMap<>();
+
+        for(SimpSession session : userSessions) {
+            Set<SimpSubscription> subs = session.getSubscriptions();
+            countSessionsFromSubs(subs, chatRoomSessions);
+        }
+
+        return chatRoomSessions;
+    }
+
+    /**
+     * @return Map with chatId as key and number of sessions as value
+     * */
+    private Map<Long, Integer> countSessionsFromSubs(Set<SimpSubscription> subs, Map<Long, Integer> buffer) {
+        for(SimpSubscription sub : subs) {
+            if(HeaderUtils.destinationStartsWith(CHAT_TOPIC_URL_PREFIX, sub.getDestination())) {
+                Long chatId = HeaderUtils.getLongFromDestination(sub.getDestination(), CHAT_ID_INDEX_IN_TOPIC_URL);
+                buffer.put(chatId, buffer.getOrDefault(chatId, 0) + 1);
+            }
+        }
+
+        return buffer;
     }
 
-    public void unsubscribe(Long chatId, String username, String sessionId, String subscriptionId) {
-        sessionContext.clearContext();
-        sessionContext.removeSubscriptionId(subscriptionId);
-        chatSessionManager.removeIfExists(chatId, username, sessionId);
+    private void sendToUser(Long chatId, String receiverUsername, String sessionId, ChatNotification notification) {
+        if(notification instanceof ChatNotificationMessage messageNotification) {
+            log.trace("Sending message notification to user: {}, session {}", receiverUsername, sessionId);
+            notifyWithMessage(chatId, receiverUsername, sessionId, messageNotification);
+        } else {
+            log.trace("Sending generic notification to user: {}, session {}", receiverUsername, sessionId);
+            notifyWithGenericMessage(chatId, receiverUsername, sessionId, notification);
+        }
     }
 
-    private void executeSendNotification(String destination, ChatNotification notification) {
-        simpMessagingTemplate.convertAndSend(destination, notification);
+    private void notifyWithMessage(
+            Long chatId,
+            String receiverUsername,
+            String sessionId,
+            ChatNotificationMessage message
+    ) {
+        ZoneId zoneId = webSocketSessionService.getTimezoneOrDefault(sessionId);
+        convertNotificationMessageTimezone(message, zoneId);
+
+        notifyWithGenericMessage(chatId, receiverUsername, sessionId, message);
+
+        if(!message.getContent().getSenderEmail().equals(receiverUsername)) {
+            chatService.updateLastMessageSeen(chatId, message.getContent().getSenderEmail());
```

Comment:
w sumie to tej metody uzywamy tylko w sendToUser, a sendToUser w sendMessages, a z kolei do wyslania wiadomości(nie powiadomienia, ze dolaczyl itp) i tak wywolujemy to tylko w controllerze, a przed wywolaniem sendMessages i tak przy tworzeniu wiadomosci oznaczamy jako przeczytana jesli odbiorca jest w czacie. Chyba, ze chodzilo o spojnosc, zeby domyslnie przy wyslaniu wiadomosci oznaczalo odbiorcy, ze ma przeczytana wiadomosc jesli jest w czacie, bo ChatSessionService przy wysylaniu nie musi zakladac, ze gdzies inna klasa z nią oznaczyla wiadomosc jako przeczytana. Ale w takim razie moze usunac oznaczanie wiadomosci jako przeczytane w controlerze

Reply:
drugiej strony to chatService.createMessage
 tez ma odpowiedzialnosc oznaczanie wiadomosci jako przeczytana, wiec moznaby usunac na poziomie tworzenia wiadomosci czy jest odczytana czy nie i dopiero przy wysylaniu oznaczac, to by chyba bylo najbardziej logiczne. Ale patrzac na to jak dzialaja websockety moze najbezpieczniej zostawic tak jak jest i podwojnie oznaczac jako przeczytane.

Reply:
To może być faktycznie do usunięcia. Możliwe że to jest pozostałość po poprzedniej wersji. Zobaczę jak w testach to wychodzi jeśli usunę tego ifa

Reply:
Chociaż ja bym bardziej chyba został nam zmiana statusu wiadomości Orzysz wysłaniu webockrtem zamiast przy tworzeniu do bazy

Reply:
Zostałem jednak przy tym zmienianiu statusu w serwisie. Tam i tak trzeba aktualizować że użytkownik zobaczył wszyskie poprzednie wiadomości jak wysłał nową więc byłaby tu rozbieżność. Jedynie w tych websocketach to ręcznie aktualizuje te statusy przy joinie